### PR TITLE
codeintel: crate bulk syncer paging

### PR DESCRIPTION
**Pull request by Strum355**
Previously, we were hitting connection timeouts and "signal: killed" a lot from gitserver, resulting in a failure to sync every crate. This PR addresses this by removing the long-held connection being used to read the entire archive by instead paging the archive in blocks of 5000 files, reading those into memory to close the connection asap. The entire archive is multiple hundred MB, which is too large for us to want to load into memory at once, hence paging in blocks of 5000 files which gives us an average of 42MB per block instead.

## Test plan

Added test case, bazinga :tada:

### CODE REVIEW

File: internal/codeintel/dependencies/internal/background/job_cratesyncer.go [hunk 1]
```diff
@@ -57,11 +61,15 @@ func NewCrateSyncer(
...
 
 	job := crateSyncerJob{
-		autoindexingSvc: autoindexingSvc,
-		dependenciesSvc: dependenciesSvc,
-		gitClient:       gitClient,
-		extSvcStore:     extSvcStore,
-		operations:      newOperations(observationCtx),
+		// average file size is ~8500bytes, 5000 files gives us an average (uncompressed) archive size of
+		// about 42MB. This will require ~21 gitserver archive calls
+		archiveWindowSize: 2000,
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/background/job_cratesyncer.go:66]: Should we make this an envvar just so we can backdoor it?

File: internal/codeintel/dependencies/internal/background/job_cratesyncer.go [hunk 2]
```diff
@@ -99,32 +107,20 @@ func (b *crateSyncerJob) handleCrateSyncer(ctx context.Context, interval time.Du
...
-	}
-	defer reader.Close()
 
-	tr := tar.NewReader(reader)
+	allFilesStr, err := j.gitClient.LsFiles(ctx, nil, repoName, "HEAD")
 	if err != nil {
 		return err
 	}
+	// safe according to rule #1 of pkg/unsafe
+	allFiles := *(*[]gitdomain.Pathspec)(unsafe.Pointer(&allFilesStr))
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/background/job_cratesyncer.go:123]: rule #1 of pkg/unsafe: we do not talk about pkg/unsafe.

File: internal/codeintel/dependencies/internal/background/job_cratesyncer.go [hunk 3]
```diff
@@ -223,6 +238,34 @@ func (b *crateSyncerJob) handleCrateSyncer(ctx context.Context, interval time.Du
...
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to git archive repo %q", repoName)
+	}
+	// important to read this into memory asap
+	defer reader.Close()
+
+	// read into mem to avoid holding connection open
+	var backing [50 * 1024 * 1024]byte
+	buf := bytes.NewBuffer(backing[:0])
```
Comment (efritz) [inline internal/codeintel/dependencies/internal/background/job_cratesyncer.go:260]: What is the difference to `buf := bytes.NewBuffer(make([]byte, 50*1024*1024))`?

Comment (efritz) [reply internal/codeintel/dependencies/internal/background/job_cratesyncer.go:260]: Or even just `io.ReadAll(reader)`?

**Review by efritz — APPROVED** [review]

File: internal/codeintel/dependencies/internal/background/job_cratesyncer.go (hunk 1 above)
Comment (efritz) [reply internal/codeintel/dependencies/internal/background/job_cratesyncer.go:66]: Yeah not a huge deal, just wondered if we'd have to tweak it

File: internal/codeintel/dependencies/internal/background/job_cratesyncer.go (hunk 3 above)
Comment (Strum355) [reply internal/codeintel/dependencies/internal/background/job_cratesyncer.go:260]: The former makes sense, the latter would (on average) result in a lot of resizing. The average filesize is about 10022bytes according to `find . -type f -printf '%s\n' | awk '{s+=$0} END {printf "Count: %u\nAverage size: %.2f\n", NR, s/NR}'`, with a window size of 5000 (changed from 2000 in the last commit) gives us an average uncompressed size of 48MB, hence the 50MB default buffer. `io.ReadAll` starts with a 512byte buffer, which would take >15 resizes

**Pull request merged**